src/snowflake/snowpark/session.py

```diff
@@ -1125,27 +1193,167 @@ def _upload_unsupported_packages(
             )
 
             # Zip and add to stage
-            zip_file = f"{IMPLICIT_ZIP_FILE_NAME}_{generate_random_alphanumeric(5)}.zip"
+            environment_signature: str = get_signature(packages)
+            zip_file = f"{IMPLICIT_ZIP_FILE_NAME}_{environment_signature}.zip"
             zip_path = os.path.join(tmpdir, zip_file)
             zip_directory_contents(target, zip_path)
 
-            # TODO: Make this upload non-lazy and allow custom stages / zip path names
-            self.add_import(zip_path)
+            # Add packages to stage
+            stage_name = self.get_session_stage()
+
+            if persist_path:
+                # Switch the stage used for storing zip file.
+                stage_name = persist_path
+
+                # Download metadata dictionary using the technique mentioned here: https://docs.snowflake.com/en/user-guide/querying-stage
+                metadata_file = f"{ENVIRONMENT_METADATA_FILE_NAME}.txt"
+                normalized_metadata_path = normalize_remote_file_or_dir(
+                    f"{persist_path}/{metadata_file}"
+                )
+                try:
+                    metadata = {
+                        row[0]: row[1] if row[1] else []
+                        for row in (
+                            self.sql(
+                                f"SELECT t.$1 as signature, t.$2 as packages from {normalized_metadata_path} t"
+                            )._internal_collect_with_tag()
+                        )
+                    }
+                except OperationalError as op_error:
+                    # The only operational error that need not be raised further, is the absence of a metadata file.
```

Comment:
can you add some unit tests to cover these lines

Reply:
Turns out if the file doesn't exist, I get an empty metadata dictionary directly, the check for `OperationalError` is not required. Removing these lines